libs/data/bbhnet/data/dataloader.py

```diff
@@ -0,0 +1,449 @@
+        Randomly sample kernels from the interferometer
+        background timeseries in a uniform manner. If
+        `independent == True`, kernels will be sampled
+        from each interferometer independently and then
+        concatenate to simulate fully random time-shifting.
+        """
+
+        # sample kernel start indices uniformly
+        # from the hanford background data ensuring
+        # that the corresponding stop index will fall
+        # within the timeseries
+        hanford_start = np.random.choice(
+            len(self.hanford_background) - self.kernel_size,
+            size=self.batch_size,
+            replace=False,
+        )
+
+        # if sampling indepdently, create a second set
+        # of start indices for the livingston background
+        # data. Otherwise, use the same indices as for hanford
+        if independent:
+            livingston_start = np.random.choice(
+                len(self.livingston_background) - self.kernel_size,
+                size=self.batch_size,
+                replace=False,
+            )
+        else:
+            livingston_start = hanford_start
+
+        # grab kernels for each start index from the background
+        # TODO: is there a good way to do this with array ops?
+        X = []
+        for h_idx, l_idx in zip(hanford_start, livingston_start):
+            hanford = self.hanford_background[h_idx : h_idx + self.kernel_size]
+            livingston = self.livingston_background[
+                l_idx : l_idx + self.kernel_size
+            ]
+            X.extend([hanford, livingston])
+
+        # stack everything to make a (batch_size * 2, sample_rate)
+        # sized tensor, then reshape to move the two
+        # interferometers to the 1st dimension
+        X = torch.stack(X, dim=0).reshape(self.batch_size, 2, -1)
+        return X
+
+    def _sample_ra(self, num_samples):
+        raise NotImplementedError
+
+    def _sample_dec(self, num_samples):
+        raise NotImplementedError
+
+    def _sample_psi(self, num_samples):
+        raise NotImplementedError
+
+    def _sample_geocent_time(self, num_samples):
+        raise NotImplementedError
+
+    def inject_waveforms(self, background, waveforms):
+        sample_params = {
+            "ra": self._sample_ra(len(background)),
+            "dec": self._sample_dec(len(background)),
+            "psi": self._sample_psi(len(background)),
+            "geocent_time": self._sample_geocent_time(len(background)),
+        }
+
+        signals = np.zeros((len(background), 2, self.kernel_size))
+        for i, ifo in enumerate(["h1", "l1"]):
+            signal = project_raw_gw(
+                waveforms,
+                sample_params,
+                self.waveform_generator,
+                ifo,
+                get_snr=False,
+            )
+            signals[:, i] = signal
+        return background + torch.Tensor(signals).to(self.device)

```

Comment:
@wbenoit26 is something like this sufficient? Or will I need to do the SNR reweighting you do [here](https://github.com/ML4GW/BBHNet/blob/9564f1595a13fa243ebc90a97b7c4025c2f6b3b6/libs/injection/bbhnet/injection/injection.py#L295-L300)? Does the fact that I do whitening by the background PSD help things at all?

Note that at this point the waveforms have been windowed into some `kernel_size` window around the trigger time. Will this affect the ifo response calculation?

Reply:
The SNR scaling will need to happen here as well if we want to have control over the range we're training on, and it might be a good idea to store the SNR regardless. Whitening will change the SNR, but I don't think that matters if all we're looking for is some consistent way to tune signal strength.

That's a good question. I think the answer is no, but I can do some digging to find out if that's true.

Reply:
Based on bilby's source code, the windowing will not affect the ifo response - it's a purely geometric quantity.

One thing that I do want to flag is the [get_detector_response](https://lscsoft.docs.ligo.org/bilby/_modules/bilby/gw/detector/interferometer.html#Interferometer.get_detector_response) function here, which I suppose is what Tri based his original injection code on. Presumably he had a reason that he needed to customize this code rather than just calling the function, but that might not be the case for us anymore.

We have a working injection script, so I don't want to keep iterating over it, but I did want to note that it exists.

Reply:
Sounds good, I'll try to get all these fixes in today so we can merge ASAP.

Good catch on the built-in function. Agreed that since what we have works we may as well keep it, but worth keeping in mind down the line.

Reply:
SNR reweighting is implemented, and the windowing ends up being a non-factor because I use the whole signal so that the fft calculation during SNR reweighting is more stable, then window it after it's been projected and reweighted